transport/src/main/java/io/netty/channel/AbstractCoalescingBufferQueue.java

```diff
@@ -221,28 +222,31 @@ public final void copyTo(AbstractCoalescingBufferQueue dest) {
      * @param ctx The context to write all elements to.
      */
     public final void writeAndRemoveAll(ChannelHandlerContext ctx) {
-        decrementReadableBytes(readableBytes);
         Throwable pending = null;
         ByteBuf previousBuf = null;
         for (;;) {
             Object entry = bufAndListenerPairs.poll();
             try {
                 if (entry == null) {
                     if (previousBuf != null) {
+                        decrementReadableBytes(previousBuf.readableBytes());
```

Comment:
If you suppose that
"while the queue is drained we may notify the promise which may add again buffers to the queue", the fix seems not correct enough.

Threre is raise condition between incrementReadableBytes
```
int nextReadableBytes = readableBytes + increment;
...
readableBytes = nextReadableBytes;
```
and decrementReadableBytes
`readableBytes -= decrement;`

Seems to me #readableBytes should be AtomicInteger.

assert readableBytes == 0; // looks cool

Reply:
@progulin maybe I am missing something but this is only meant to be used by a single thread. What race you are talking about ?

Reply:
I supposed writeAndRemoveAll/releaseAndCompleteAll is one thread and another which does add/addFirst.
If single thread is used, there is no sence to duplicate decrementReadableBytes() calls.
Unpached version set readableBytes to zero `decrementReadableBytes(readableBytes);` and clear queue into cycle `for (;;)`. The same does patched.

Reply:
@progulin yes all is the single threaded... Sorry I have no idea what you are talking about. The problem is the promise may be notified and so the `ChannelFutureListener` may call `add(...)` while we are in the for loop. All of this is in the same thread. Just run the unit test with the unpatched version and you will see it fails.

Reply:
> ChannelFutureListener may call add(...)
Ugu! Now I see. Thank you. Great patch.